From 1f6788321a240edb83e018619e0392fc53d0c6e1 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 17 Jun 2021 18:46:48 +0200
Subject: [PATCH] PONRTSYS-9036: Fix modifying VLAN priority to 0

Instead of assuming that VLAN priority 0 means that we should use the
priority of the inner or outer tag, check if the VLAN priority was set
in the action. This fixes rules which explicitly modify the VLAN
priority to 0.

This fixes some scrambled version of BBF 6.1.34.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
index 299609adbf38..92bc9bd81971 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
@@ -278,10 +278,12 @@ static void vlan_action_modify_prio_parse(struct net_device *dev,
 					  enum tc_flower_vlan_tag tag,
 					  int *prio)
 {
-	*prio = tcf_vlan_push_prio(a);
+	u8 flags = tcf_vlan_push_flags(a);
 
-	if (*prio)
+	if (flags & ACTVLAN_PUSH_F_PRIO) {
+		*prio = tcf_vlan_push_prio(a);
 		return;
+	}
 
 	if (tag == TC_VLAN_SINGLE_TAGGED)
 		*prio = CP_FROM_INNER;
